packages/gatsby/src/utils/webpack/cache-folder-resolver.ts

```diff
@@ -0,0 +1,76 @@
+import Resolver from "enhanced-resolve/lib/Resolver"
+
+interface IRequest {
+  request?: string
+  path: string
+}
+
+type ProcessWithPNP = NodeJS.ProcessVersions & { pnp?: string }
+
+/**
+ * To support PNP we have to make sure dependencies resolved from the .cache folder should be resolved from the gatsby package directory
+ */
+export class CacheFolderResolver {
+  private requestsFolder: string
+
+  constructor(requestsFolder: string) {
+    this.requestsFolder = requestsFolder
+  }
+
+  apply(resolver: Resolver): void {
+    if (!(process.versions as ProcessWithPNP).pnp) {
+      return
+    }
+
+    const target = resolver.ensureHook(`raw-module`)
+    resolver
+      .getHook(`raw-module`)
+      .tapAsync(
+        `CacheFolderResolver`,
+        (
+          request: IRequest,
+          resolveContext: unknown,
+          callback: (err?: Error | null, result?: unknown) => void
+        ) => {
+          const req = request.request
+          if (!req) {
+            return callback()
+          }
+
+          if (!request.path.startsWith(this.requestsFolder)) {
+            return callback()
+          }
+
+          const packageMatch = /^(@[^/]+\/)?[^/]+/.exec(req)
```

Comment:
won't this also match requests like `./foo`? or they are already absolute at this point?

Reply:
Normally it would but because we're listening on raw-module we only get real modules (I don't know how to say it, modules from node_modules). 

```js
		new ConditionalPlugin(
			"after-normal-resolve",
			{ module: true },
			"resolve as module",
			false,
			"raw-module"
		)
```
https://github.com/webpack/enhanced-resolve/blob/60d79f3c93304ce5ecbbe0127aa583d4a73bf1a1/lib/ResolverFactory.js#L359-L365